### PR TITLE
Added smart server list

### PR DESCRIPTION
**Pull request by Netshroud**
- Remember which servers we have recently managed to connect to, and which servers we have recently failed to connect to. Consumers can customise the definition of "recently".
- Prefer to reconnect to a known good server than a neutral or known bad server.
- Prefer to reconnect to a neutral server than a known bad server.
- Added API for loading the server list from Steam Directory Web API. Default (no consumer action) continues to use the hardcoded list of IPEndPoints.
- Hooked server list into CMClient to automatically detect good and bad servers.
- Added sample illustrating basic usage of using the server list API to persist the server list between executions.

### CODE REVIEW

File: SteamKit2/SteamKit2/Networking/Steam3/CMServerList.cs [hunk 1]
```diff
@@ -0,0 +1,361 @@
...
+                if ( result != EResult.OK )
+                {
+                    throw new InvalidOperationException( string.Format( "Steam Web API returned EResult.{0}", Enum.GetName( typeof( EResult ), result ) ) );
+                }
+
+                var list = response[ "serverlist" ];
+
+                cancellationToken.ThrowIfCancellationRequested();
+
+                lock (listLock )
```
Comment (xPaw) [inline SteamKit2/SteamKit2/Networking/Steam3/CMServerList.cs]: Holy whitespace batman!

File: SteamKit2/SteamKit2/Networking/Steam3/CMServerList.cs [hunk 2]
```diff
@@ -0,0 +1,361 @@
...
+        /// Load a list of servers from the Steam Directory
+        /// </summary>
+        /// <param name="cellid">Cell ID</param>
+        public Task LoadListFromDirectoryAsync( int cellid = 0 )
+        {
+            return LoadListFromDirectoryAsync( cellid, CancellationToken.None );
+        }
+
+        /// <summary>
+        /// Load a list of servers from the Steam Directory
```
Comment (xPaw) [inline SteamKit2/SteamKit2/Networking/Steam3/CMServerList.cs]: This should have a better description of what it does, aka it should say that it's going to replace current servers, and should be used once before initial connection.

File: SteamKit2/SteamKit2/Networking/Steam3/CMServerList.cs [hunk 3]
```diff
@@ -0,0 +1,360 @@
...
+            cancellationToken.ThrowIfCancellationRequested();
+
+            var task = directory.Call( "GetCMList", version: 1, args: args, secure: true );
+            return task.ContinueWith(t =>
+            {
+                var response = task.Result;
+                var result = ( EResult )response[ "result" ].AsInteger( ( int ) EResult.Invalid );
+                if ( result != EResult.OK )
+                {
+                    throw new InvalidOperationException( string.Format( "Steam Web API returned EResult.{0}", Enum.GetName( typeof( EResult ), result ) ) );
```
Comment (VoiDeD) [inline SteamKit2/SteamKit2/Networking/Steam3/CMServerList.cs]: Whitespace, and is Enum.GetName needed here?

File: SteamKit2/SteamKit2/Networking/Steam3/CMServerList.cs [hunk 4]
```diff
@@ -0,0 +1,360 @@
...
+        {
+            serverInfo.Score = score;
+            serverInfo.LastScoreChangeTimeUtc = DateTime.UtcNow;
+        }
+
+        /// <summary>
+        /// Get the next server in the list.
+        /// </summary>
+        /// <returns>An <see cref="System.Net.IPEndPoint"/>, or null if the list is empty.</returns>
+        public IPEndPoint GetNextServer()
```
Comment (VoiDeD) [inline SteamKit2/SteamKit2/Networking/Steam3/CMServerList.cs]: `GetIdealServer` perhaps?

File: SteamKit2/SteamKit2/Steam/CMClient.cs [hunk 5]
```diff
@@ -94,43 +95,8 @@ public IPAddress LocalIP
...
-                new IPEndPoint( IPAddress.Parse( "208.78.164.12" ), 27019 ),
-                new IPEndPoint( IPAddress.Parse( "208.78.164.13" ), 27017 ),
-                new IPEndPoint( IPAddress.Parse( "208.78.164.13" ), 27018 ),
-                new IPEndPoint( IPAddress.Parse( "208.78.164.13" ), 27019 ),
-                new IPEndPoint( IPAddress.Parse( "208.78.164.14" ), 27017 ),
-                new IPEndPoint( IPAddress.Parse( "208.78.164.14" ), 27018 ),
-                new IPEndPoint( IPAddress.Parse( "208.78.164.14" ), 27019 ),
-            } );
+            Servers = new SmartCMServerList();
+            Servers.UseInbuiltList();
```
Comment (VoiDeD) [inline SteamKit2/SteamKit2/Steam/CMClient.cs]: Nit: if we're always doing this, may as well throw it into the constructor

Comment (Netshroud) [inline SteamKit2/SteamKit2/Steam/CMClient.cs]: I'd prefer not to have every test start with `Clear()`. We're only doing this once though technically - it's in a static constructor.

**Comment by Netshroud** [comment]
OK, I think that's everything now.

**Comment by xPaw** [comment]
:+1:

**Comment by Netshroud** [comment]
Last call for comments before :ship:ing time.

File: SteamKit2/SteamKit2/Networking/Steam3/CMServerList.cs [hunk 6]
```diff
@@ -0,0 +1,334 @@
+﻿using System;
+using System.Collections.Generic;
+using System.Collections.ObjectModel;
+using System.Diagnostics;
+using System.Linq;
+using System.Net;
+using System.Threading;
+using System.Threading.Tasks;
+
+namespace SteamKit2.Networking.Steam3
```
Comment (VoiDeD) [inline SteamKit2/SteamKit2/Networking/Steam3/CMServerList.cs]: One last bit that's been bothering me. Do we want to introduce this namespace? I've always viewed the `Networking` subdirectory as something for the internal bits for connections, whereas managing a server list is something that lives on top of `CMClient`.

Comment (Netshroud) [inline SteamKit2/SteamKit2/Networking/Steam3/CMServerList.cs]: Good catch!

**Pull request merged**